src/v/utils/truncating_logger.h

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "seastarx.h"
+#include "ssx/sformat.h"
+#include "vassert.h"
+
+#include <seastar/util/log-impl.hh>
+#include <seastar/util/log.hh>
+
+#include <iostream>
+
+/// Logger that supports proactive truncation of formatted output at the level
+/// of seastar::logger::log_writer
+class truncating_logger {
+    static constexpr std::string_view trunc_msg_fmt
+      = "... log line too large; dropped {}B";
+    static constexpr size_t max_max_line_bytes = 1048576;
+    static constexpr size_t max_trunc_msg_len = trunc_msg_fmt.size() + (20 - 2);
+
+public:
+    truncating_logger(ss::logger& logger, size_t max_line_bytes)
+      : _logger(logger) {
+        vassert(
+          max_line_bytes <= max_max_line_bytes,
+          "max_line_bytes {} too large; should be <= {}",
+          max_line_bytes,
+          max_max_line_bytes);
+        vassert(
+          max_line_bytes > max_trunc_msg_len,
+          "max_line_bytes {} too small; should be > {}",
+          max_line_bytes,
+          max_trunc_msg_len);
+
+        _max_line_bytes = max_line_bytes - max_trunc_msg_len;
+    }
+
+    template<typename... Args>
+    void log(ss::log_level lvl, const char* format, Args&&... args) const {
+        if (_logger.is_enabled(lvl)) {
+            ss::logger::lambda_log_writer writer(
+              [&](ss::internal::log_buf::inserter_iterator it) {
+                  auto res = fmt::format_to_n(
+                    it,
+                    _max_line_bytes,
+                    fmt::runtime(format),
```

Comment:
Is figuring out the compile-time checks is worthwhile, here?

Reply:
Which compile time checks?

Reply:
> Which compile time checks?

[format_string](https://fmt.dev/8.1.1/api.html#c.fmt:~:text=format_string%20is%20a%20format%20string%20which%20can%20be%20implicitly%20constructed%20from%20a%20string%20literal%20or%20a%20constexpr%20string%20and%20is%20checked%20at%20compile%20time%20in%20C%2B%2B20.%20To%20pass%20a%20runtime%20format%20string%20wrap%20it%20in%20fmt%3A%3Aruntime().)

Reply:
Did you get a chance to look at this?

Reply:
ah right; not yet. attention has been somewhat split. will do